src/vm/node_modules/VM.js

```diff
@@ -10957,6 +10958,384 @@ exports.receive = function (target, options, callback)
+                        if (vmobj.disks[i].boot !== true) {
+                            next(new Error('disk0 found but does not have '
+                                + 'boot property set to true - reprovision '
+                                + 'requires a bootable disk0'));
+                            return;
+                        }
+                        disk0 = vmobj.disks[i];
+                        disk0_ds = ds;
+                        disk0_path = vmobj.disks[i].path;
+                        break;
+                    }
+                }
+            }
+
+            if (!disk0) {
+                next(new Error('disk0 not found for bhyve VM'));
+                return;
+            }
+
+            log.info('Found disk0 at %s (dataset: %s)',
+                disk0_path, disk0_ds);
+            next();
+        },
+        function findSnapshot(next) {
+            var snapname = vmobj.zpool + '/' + payload.image_uuid + '@final';
+
+            // Check if snapshot exists
+            zfs(['get', '-Ho', 'value', 'type', snapname], log,
+                function (err, fds) {
+                    if (!err) {
+                        log.debug('Using snapshot: %s', snapname);
+                        snapshot = snapname;
+                        next();
+                        return;
+                    }
+
+                    if (fds.stderr.match(/dataset does not exist/)) {
+                        // Try VM-specific snapshot
+                        snapname = vmobj.zpool + '/' + payload.image_uuid
+                            + '@' + vmobj.uuid;
+
+                        zfs(['get', '-Ho', 'value', 'type', snapname], log,
+                            function (err2, fds2) {
+                                if (!err2) {
+                                    log.debug('Using snapshot: %s', snapname);
+                                    snapshot = snapname;
+                                    next();
+                                } else if (fds2.stderr.match(
+                                    /dataset does not exist/)) {
+                                    // Cannot create snapshot - image dataset
+                                    // doesn't exist
+                                    next(new Error('Image ' + payload.image_uuid
+                                        + ' not found in zpool '
+                                        + vmobj.zpool));
+                                } else {
+                                    next(err2);
+                                }
+                            });
+                    } else {
+                        next(err);
+                    }
+                });
+        },
+        function destroyOldDisk0(next) {
+            // Destroy the old disk0 - now that we've verified the
+            // snapshot exists
+            log.info('Destroying old disk0: %s', disk0_ds);
+            zfs(['destroy', '-f', disk0_ds], log, function (err, fds) {
```

Comment:
You will most likely need to add `'-r'` to the arguments.  See OS-2270 (and friends) and https://github.com/TritonDataCenter/smartos-live/commit/4bc527d46d1245959b8c4df8d79ef59b1c59d3af for why.

Reply:
1.) There might be a flag day on the sender side too.  No idea what revisions you're using on either end.  Probably not the issue.

2.) I tested this during OS-8518, but there may be other send/recv problems relating to refreservation and other issues.  I highly recommend using -r anyway.

Reply:
A test would be: for both a native zone pre-and-post-this-fix, and post-this-fix a bhyve zone:

`zfs snapshot -r zones@test`

`vmadm reprovision...`

Check success/failure.

Reply:
Should be fixed in https://github.com/TritonDataCenter/smartos-live/pull/1156/commits/dd45e25bc110e428548fa23bb74e210b3d23a766

Flowchart, tests, and test results updated as well.